File: backend/app/schemas/comparison.py
"""
Schemas Pydantic para o módulo de auditoria TR × Propostas.

Inclui fornecedores, moldes de regras e resultados de comparação.
"""

import uuid
from datetime import datetime, timezone
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, BeforeValidator


def _ensure_tz(v: datetime) -> datetime:
    if isinstance(v, datetime) and v.tzinfo is None:
        return v.replace(tzinfo=timezone.utc)
    return v


AwareDatetime = Annotated[datetime, BeforeValidator(_ensure_tz)]


class FornecedorCreate(BaseModel):
    """Dados para cadastrar um fornecedor."""
    nome: str = Field(..., max_length=500)
    cnpj: str | None = Field(default=None, max_length=18)
    email: str | None = Field(default=None, max_length=255)


class FornecedorResponse(BaseModel):
    """Resposta com dados de um fornecedor."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nome: str
    cnpj: str | None = None
    email: str | None = None
    created_at: AwareDatetime


class FornecedorListResponse(BaseModel):
    """Lista de fornecedores."""
    fornecedores: list[FornecedorResponse]
    total: int


class MoldeCreate(BaseModel):
    """Dados para criar um molde de regras."""
    nome: str = Field(..., max_length=200)
    descricao: str | None = None
    config_json: str


class MoldeResponse(BaseModel):
    """Resposta com dados de um molde."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nome: str
    descricao: str | None = None
    config_json: str
    created_at: AwareDatetime


class MoldeListResponse(BaseModel):
    """Lista de moldes."""
    moldes: list[MoldeResponse]
    total: int


class ComparacaoResultadoResponse(BaseModel):
    """Resultado de uma regra para um fornecedor."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    fornecedor_id: uuid.UUID
    regra_id: str
    status: str
    motivo: str | None = None
    valor_tr: str | None = None
    valor_proposta: str | None = None


class ComparacaoStartResponse(BaseModel):
    """Resposta ao iniciar uma comparação."""
    comparacao_id: uuid.UUID
    message: str


class ComparacaoResponse(BaseModel):
    """Resposta resumida de uma comparação."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tr_document_id: uuid.UUID
    molde_id: uuid.UUID
    status: str
    error_message: str | None = None
    created_at: AwareDatetime
    completed_at: AwareDatetime | None = None
    total_resultados: int = 0
    fornecedores: list[FornecedorResponse] = []


class MatrizResponse(BaseModel):
    """Matriz de conformidade: regras × fornecedores."""
    comparacao_id: uuid.UUID
    tr_document_id: uuid.UUID
    status: str
    regras: list[str]
    fornecedores: list[FornecedorResponse]
    linhas: list["MatrizLinha"]


class MatrizLinha(BaseModel):
    """Uma linha da matriz (uma regra para todos os fornecedores)."""
    regra_id: str
    rotulo: str
    celulas: list["MatrizCelula"]


class MatrizCelula(BaseModel):
    """Uma célula da matriz: status da regra para um fornecedor."""
    fornecedor_id: uuid.UUID
    status: str
    motivo: str | None = None
    valor_tr: str | None = None
    valor_proposta: str | None = None


class FeedbackFalha(BaseModel):
    """Falha no envio de e-mail para um fornecedor."""
    fornecedor_id: uuid.UUID
    nome: str
    email: str | None = None
    motivo: str


class FeedbackResponse(BaseModel):
    """Resumo do envio de pendências por e-mail (RF04)."""
    comparacao_id: uuid.UUID
    enviados: int
    falhas: list[FeedbackFalha] = []
    fornecedores_sem_pendencias: list[str] = []
    fornecedores_sem_email: list[str] = []

File: frontend/src/types/comparison.ts
/**
 * Tipos de auditoria TR × Propostas (fornecedores, moldes, matriz) — extraído de `types/index.ts`.
 */

export type ComparacaoStatus = 'pending' | 'running' | 'completed' | 'error';

export type ConformidadeStatus = 'ok' | 'falha' | 'atencao';

export type AnchorTipo =
  | 'numero_inteiro'
  | 'numero_extenso'
  | 'booleano'
  | 'legal'
  | 'data'
  | 'percentual'
  | 'monetario'
  | 'cnpj'
  | 'prazo_relativo'
  | 'cep';

export interface Fornecedor {
  id: string;
  nome: string;
  cnpj: string | null;
  email: string | null;
  created_at: string;
}

export interface FornecedorListResponse {
  fornecedores: Fornecedor[];
  total: number;
}

export interface Molde {
  id: string;
  nome: string;
  descricao: string | null;
  config_json: string;
  created_at: string;
}

export interface RegraConfig {
  id: string;
  rotulo: string;
  tipo: AnchorTipo;
  ancora?: string | null;
  unidade?: string | null;
  expectativa?: number | string | null;
  palavras_chave?: string[] | null;
  regex?: string | null;
}

export interface MoldeConfig {
  versao: number;
  regras: RegraConfig[];
}

export interface MoldeListResponse {
  moldes: Molde[];
  total: number;
}

export interface ComparacaoStartResponse {
  comparacao_id: string;
  message: string;
}

export interface ComparacaoResponse {
  id: string;
  tr_document_id: string;
  molde_id: string;
  status: ComparacaoStatus;
  error_message: string | null;
  created_at: string;
  completed_at: string | null;
  total_resultados: number;
  fornecedores: Fornecedor[];
}

export interface ComparacaoListResponse {
  comparacoes: ComparacaoResponse[];
  total: number;
}

export interface MatrizCelula {
  fornecedor_id: string;
  status: ConformidadeStatus;
  motivo: string | null;
  valor_tr: string | null;
  valor_proposta: string | null;
}

export interface MatrizLinha {
  regra_id: string;
  rotulo: string;
  celulas: MatrizCelula[];
}

export interface MatrizResponse {
  comparacao_id: string;
  tr_document_id: string;
  status: string;
  regras: string[];
  fornecedores: Fornecedor[];
  linhas: MatrizLinha[];
}

// ---- Dry-Run de moldes ----

export interface DryRunResultado {
  regra_id: string;
  rotulo: string;
  tipo: AnchorTipo;
  ancora: string | null;
  valor_extraido: string | null;
  encontrado: boolean;
}

export interface DryRunResponse {
  molde_id: string;
  molde_nome: string;
  documento_id: string;
  documento_nome: string;
  total_regras: number;
  regras_encontradas: number;
  resultados: DryRunResultado[];
}

export const COMPARACAO_STATUS_LABELS: Record<ComparacaoStatus, string> = {
  pending: 'Pendente',
  running: 'Comparando...',
  completed: 'Concluído',
  error: 'Erro',
};

export const CONFORMIDADE_LABELS: Record<ConformidadeStatus, string> = {
  ok: 'OK',
  falha: 'FALHA',
  atencao: 'ATENÇÃO',
};

export const ANCHOR_TIPO_LABELS: Record<AnchorTipo, string> = {
  numero_inteiro: 'Número inteiro',
  numero_extenso: 'Número por extenso',
  booleano: 'Booleano (presença)',
  legal: 'Referência legal (regex)',
  data: 'Data',
  percentual: 'Percentual',
  monetario: 'Valor monetário (R$)',
  cnpj: 'CNPJ',
  prazo_relativo: 'Prazo relativo (ex: 30 dias)',
  cep: 'CEP',
};
